### ISSUE TITLE
Netlink elements ?

### ISSUE THREAD

**Issue by sebbbastien** [gabstopper/smc-python#1]
Hi,
I can not figure out how to create static-netlinks through smc-python APIs.
The documentation available on ReadTheDoc does not mention this type of element.

Can you help me ?

**Comment by gabstopper** [comment]
Hi Sebastien,
Thanks for the email. You are correct on Netlink's, they have not yet been
implemented which is why the documentation is not there. I will have that
implemented this weekend and will update the issue on github, along with
examples. I'm assuming you will need to do the following: create the
netlink, add to multilink, attach to the engine interface? I will also add
example of adding a multilink to a NAT rule in case that is required as
well. Thanks for flagging this. Do you have any feedback on other aspects
of the library?

David

**Comment by sebbbastien** [comment]
Hi,
I don't use multilink. I need static netlink for MPLS multi-homing (a cluster connected to a sinngle MPLS with two distinct interconnections :+1: 
![image](https://cloud.githubusercontent.com/assets/5331611/25305619/2121edea-277f-11e7-8440-c4b1e5a92597.png)

An other need I can't do today is to search network based on ip/netmask (or ip/cidr). If I search 10.0.0.0 in network objects, I'll get 10.0.0.0/8 and 10.0.0.0/12 and 10.0.0.0/29. It can be usefull to search networks with ip/cidr.

I use this library for only a few days, and for now it very well answers to our needs (provisioning configurations for many remote sites).

Best regards,

**Comment by gabstopper** [comment]
Hi Sebastien, 
Commit d557817 adds support for static netlinks. 
You can add your netlinks by code like this:

```python
from smc.elements.netlink import StaticNetlink

StaticNetlink.create(......)

engine = Engine('vm')
rnode = engine.routing.get('0.133')  #shortcut to get routing node by interface id
rnode.add_traffic_handler(
    netlink=StaticNetlink('mynetlink'),
    netlink_gw=Router('myrouter'))
```

Documentation for netlinks are under engine->routing->traffic handlers:
http://smc-python.readthedocs.io/en/latest/pages/reference.html#module-smc.elements.netlink

For searching by /cidr, the SMC API reserves the '/' char which means searches will truncate at this character. You will see the same thing when searching in the SMC UI as well. However for now, you could implement this workaround:

```python
def exact_network(network, with_cidr):
    networks = list(Network.objects.filter(network))
    for network in networks:
        if network.ipv4_network == with_cidr:
            return network
```
(I added ipv4_network and ipv6_network properties to the Network class).

Call it like this to get the first match from the filtered list:

```python
exact_network('172.18.1.0', '172.18.1.0/24')
```
The Network.objects.filter() call generates only one query and returns base meta data only based on a search query. The call to ``network.ipv4_network`` requires an additional query per element in the list to retrieve all attributes for that network to find the exact match. The function will short circuit out once it finds it's first match.

Let me know if this works for you.

**Comment by sebbbastien** [comment]
Hi,

I adapt your workaround like this:
`from ipaddress import ip_address, ip_network
def get_network(network_with_cidr):
    network = ip_network(network_with_cidr, strict=False)

    networks = list(Network.objects.filter(network.network_address.exploded))
    for net in networks:
        if net.ipv4_network == network.exploded:
            return net
            
    return False`

It works  perfectly for me.

I'll try using Netlink this afternoon.

Many thank you for your help !

**Comment by gabstopper** [comment]
Sounds good. Thanks for sharing. I forgot to mention that if you need to remove a netlink, you can also do so by (remove static netlink named 'netlink' from interface 0 :
```python
rnode = engine.routing.get(0)
rnode.remove_route_element(StaticNetlink('netlink'))
```
This will also remove the nested group/etc from under the netlink.

**Comment by gabstopper** [comment]
Did you have a chance to validate Static Netlink's working for you?

**Comment by sebbbastien** [comment]
Hi,
I'm actually stuck on objects manipulation.

I need to deal with an existing SMC with thousand of objects, with no naming convention.

Each time I need one object, I have to check if it already exists, or if I have to create it.

It is very simple to get an object from his name, but doing it on his value is pretty difficult and redondant.

I wrote thoses lines to create a network or update the name of an existing one: 

```python
In [2]: def get_network(network_with_cidr):
   ...:     network = ip_network(network_with_cidr, strict=False)
   ...: 
   ...:     networks = list(Network.objects.filter(network.network_address.exploded))
   ...:     for net in networks:
   ...:         if net.ipv4_network == network.exploded:
   ...:             return net
   ...: 
   ...:     return False
   ...: 
   ...: def update_network(net, net_name):
   ...:     net.data['name'] = net_name
   ...:     net.update()
   ...: 
   ...: def create_network(net_name, net_addr):
   ...:     Network.create(net_name, net_addr)
   ...: 
   ...: def create_or_update_net(net_name, net_addr):
   ...:     net = get_network(net_addr)
   ...: 
   ...:     if net: update_network(net, net_name)
   ...:     else: create_network(net_name, net_addr)
   ...:     

In [3]: Network.create("net-a", "2.3.4.0/24")
Out[3]: 'http://172.31.254.14:8082/6.1/elements/network/1194'

In [4]: create_or_update_net("net-b", "2.3.4.0/24")

In [5]: Network("net-b").data
Out[5]: 
{'broadcast': False,
 'ipv4_network': '2.3.4.0/24',
 'key': 1194,
 'link': [{'href': 'http://172.31.254.14:8082/6.1/elements/network/1194',
   'method': 'GET',
   'rel': 'self',
   'type': 'network'},
  {'href': 'http://172.31.254.14:8082/6.1/elements/network/1194/export',
   'method': 'POST',
   'rel': 'export'},
  {'href': 'http://172.31.254.14:8082/6.1/elements/network/1194/search_category_tags_from_element',
   'method': 'GET',
   'rel': 'search_category_tags_from_element'}],
 'name': 'net-b',
 'read_only': False,
 'system': False}
```

To configure Static Netlinks I also need Routers objects. Maybe there is already one with the good IP but a wrong name. So I have to write the same kind of functions to create or update les objets.

Am I missing something? Do I use correctly thoses REST APIs.

**Comment by gabstopper** [comment]
Hi,
Yes, it looks like you are using it properly. I can include helper functions like this for simple object creation to simplify this as well. Part of the initial hesitation is that this works for simple elements which may only require a simple value (i.e. Network just needs name and ipv4_network at minimum). Others may be more complex and require additional settings.

When searching, there are other shortcuts that may be useful. All elements are searchable by their 'name', 'comment' and primary setting field. By primary setting field I mean for Network it is 'ipv4_network' or 'ipv6_network' values, for Hosts by 'address' field, etc. 

So you can do searches like this (if you don't know the name but know the value you want to use:

```python
>>> list(Router.objects.filter('172.18.6.241'))
[Router(name=router-172.18.6.241)]

>>> from smc.elements.network import Host
>>> Host.create('hax0r', address='1.1.1.1', comment='foobar')
'http://172.18.1.150:8082/6.2/elements/host/104274'
...
>>> list(Host.objects.filter('foobar'))   # <-- by comment field
[Host(name=hax0r)]
>>>
>>> list(Host.objects.filter('1.1.1.1'))  # <-- by address field
[Host(name=hax0r), Host(name=host), Host(name=tylercharles), Host(name=hostelement), Host(name=abcdefghijklmnop)]
>>> list(Host.objects.filter('hax*'))  # <-- partial name match. Use * for wildcard
[Host(name=hax0r)]
```

There are also context filters that allow you to more broadly search across elements by categories and use the same syntax and filtering logic.

Context filters are: 
'fw_clusters', 'engine_clusters', 'ips_clusters', 'layer2_clusters', 'network_elements', 'services', 'services_and_applications', 'tags', 'situations'

So for example, finding a service with port 8080:

```python
>>> from smc.base.collection import Search
>>> list(Search('services').objects.filter('8080'))
[Protocol(name=HTTP8080), TCPService(name=TCP_8080)]
```

Same for other types:
```python
>>> list(Search('network_elements').objects.filter('1.1.1.1'))
[Host(name=hax0r), DNSRelayProfile(name=mynewprofile), Host(name=host), Host(name=tylercharles), AddressRange(name=rangeelement), IPS(name=ips), Host(name=hostelement), AddressRange(name=myrange), Host(name=abcdefghijklmnop), Router(name=myrouter), AddressRange(name=tyler), AddressRange(name=test2)]
```

You can also chain specific object filters together to find a subset of elements:
```python
>>> list(Search('host,router').objects.filter('1.1.1.1'))
[Host(name=hax0r), Host(name=host), Host(name=tylercharles), Host(name=hostelement), Host(name=abcdefghijklmnop), Router(name=myrouter)]
```

All available filters are based on the SMC API entry points available. Entry points are direct links into the resource. To find all available filters used for Search, you can run:
```python
>>> Search.object_types()
```

If you have specific element types that you might want to categorize, you might consider using a CategoryTag on the element, and then you have 'containers' for your element searches as well.
```python
>>> from smc.elements.other import CategoryTag
>>> CategoryTag.create(name='myelements', comment='container for my elements')
'http://172.18.1.150:8082/6.2/elements/category_tag/3566'
>>> 
>>> c = CategoryTag('myelements')
>>> c.add_element(Host('kali'))
>>> h = Host('kali')
>>> h.category_tags
[CategoryTag(name=myelements), CategoryTag(name=foocategory)]
```
Then use CategoryTag context searches:
```python
>>> list(CategoryTag.objects.filter('myelements'))
[CategoryTag(name=myelements)]
>>> for x in list(CategoryTag.objects.filter('myelements')):
...   x.search_elements()
... 
[Host(name=kali)]
>>> 
```

If you find you think about searches in a different way, any feedback would be appreciated, i'm happy to improve this area as needed.

**Comment by gabstopper** [comment]
Hi,
I forgot to mention, when I look at your example above you are doing this:
```python
  def update_network(net, net_name):
      net.data['name'] = net_name
      net.update()
```
There is another more generic modifier method that all classes of type ``Element`` inherit called modify_attribute. This will call update() internally. It might be a bit easier to handle. Instance cache is also flushed like when calling .update():
```python
>>> h = Host('blahtest')
>>> h.modify_attribute(name='newhost')
>>> h.data
{'name': 'newhost', 'key': 18773, 'link': [{'rel': 'self', 'href': 'http://172.18.1.150:8082/6.2/elements/host/18773', 'type': 'host'}, {'rel': 'export', 'href': 'http://172.18.1.150:8082/6.2/elements/host/18773/export'}, {'rel': 'search_category_tags_from_element', 'href': 'http://172.18.1.150:8082/6.2/elements/host/18773/search_category_tags_from_element'}], 'read_only': False, 'system': False, 'secondary': [], 'address': '23.23.23.23', 'third_party_monitoring': {'netflow': False, 'snmp_trap': False}}
```

This is basically just a generic modifier. 
If I can understand your use case, I can add properties to simplify having to modify the attributes directly also. Based on what I understand, I will add common methods to class ``Element`` that will span all element types such as 'rename', 'comment', 'add_category_tag'.

**Comment by sebbbastien** [comment]
Hi,

Sorry for the delay, a lot of work beside...

> If you find you think about searches in a different way, any feedback would be appreciated, i'm happy to improve this area as needed.

Searches on primary settings are very usefull, but I think they need more options, like regex, or at least strict/loose search.

If I need to get `1.1.1.1`, current search will return `11.1.1.1` and `1.1.1.11`. Using regex would be perfect, but if I can at least use request with strict match it will be fine.

**Comment by gabstopper** [comment]
Hi there, no problem, I know how it goes.. 

Thanks for the idea on regex, i'll think through that. I do already have strict match implemented. You just add an 'exact_match=True' to your filter query:
```python
>>> from smc.elements.network import Host
>>> list(Host.objects.filter('kali'))
[Host(name=kali), Host(name=kali2), Host(name=kali3)]
>>> list(Host.objects.filter('kali', exact_match=True))
[Host(name=kali)]
>>> 
```

**Comment by gabstopper** [comment]
You can also use a wildcard '*' when searching as well:
```python
>>> list(Host.objects.filter('ka*'))
[Host(name=kali), Host(name=kali2), Host(name=kali3)]
```

**Comment by sebbbastien** [comment]
Hi,

It seems that `exact_match` works only with object names:

```python
In [2]: from smc.elements.network import Router

In [3]: Router.create('R1', address="10.10.10.1")

In [4]: Router.create('R2', address="10.10.10.11")

In [5]: Router.create('R3', address="110.10.10.1")

In [6]: list(Router.objects.filter('10.10.10.1'))
Out[6]: [Router(name=R1), Router(name=R2), Router(name=R3)]

In [7]: list(Router.objects.filter('10.10.10.1', exact_match=True))
Out[7]: [Router(name=R1), Router(name=R2), Router(name=R3)]
```

Is it hard to change this behavior?

**Comment by gabstopper** [comment]
Hi, this is being returned by the SMC framework, i'll ping dev and see what options we have. I see the same results here for non-name fields and exact_match. I have an idea on addressing in smc-python but would rather see it fixed in the SMC API if possible.

**Comment by gabstopper** [comment]
Hi there, after reviewing on the SMC API side, I will make enhancements to the search capabilities on the smc-python client side to incorporate the ability to provide kwargs to filter searches based on a specific field. In the syntax:
Network.objects.filter(address=10.10.10.1) (for example). This will then use a generic search to find the network element using '10.10.10.1' and further filter the responses based on retrieving each result set to make a more exact match.
Also, I will implement get_or_create logic on the smc.base.model.Element type to allow you to implement the logic in form:

host = Host.get_or_create(name='myhost', address='12.12.12.12'). 
This will result in a query to find the host by specified address and return that element, or create. Will re-post when complete.

**Comment by gabstopper** [comment]
Hi,
Sorry for the delay on this one. 

I've posted a new update that has some additional features that should help you with some of the boilerplate code:

Added get_or_create to the smc.base.Element class:

```
Host.get_or_create(
    filter_key={'address': '1.1.1.5'},
    name='mynewhost',
    address='1.1.1.1',
    secondary=['1.1.1.2']))
```
``filter_key`` is an optional parameter where you can specify a key/value pair for the element attribute and value to match. In the example above, a search is first done for 1.1.1.5 in the address field of Host elements. If it exists, it is returned. If not, then one is created with the remaining kwargs. Remaining kwargs are based on each elements ``create()`` constructor.

Searching for network elements should be much easier, i've implemented a 'filter_key' optional filter on collections (amongst other collection based methods). This will help get a more exact match by interrogating the elements returned from a search.

Using your router example from earlier:
```
>>> iterator = Router.objects.iterator()
>>> list(iterator.filter('10.10.10.1'))
[Router(name=Router-110.10.10.10), Router(name=Router-10.10.10.10), Router(name=Router-10.10.10.1)]
>>> list(iterator.filter('10.10.10.1').filter_key(['address']))
[Router(name=Router-10.10.10.1)]
```

In the example above, 'filter_key' is used to specify an attribute to match 'filter' against. This will solve the 'contains' searches that the SMC performs.

For your network element matching, the same applies, but now the match is done internally within the collection. I've escaped out metacharacters that are reserved in SMC (forward slash and dash):

```
>>> list(Network.objects.filter('172.18.1.0/24').filter_key(['ipv4_network']))
[Network(name=internal-network), Network(name=net-172.18.1.0/24), Network(name=network-172.18.1.0/24)]
>>> Network.objects.filter('172.18.1.0/24').filter_key(['ipv4_network']).first()
Network(name=internal-network)
```
From the above, I added methods ``first``, ``last``, ``count`` and ``exists`` to the collection. So instead of parsing the list results or checking for results, you have some other choices:

For example, checking for existence of results and then obtaining:
```
>>> iterator = Network.objects.iterator()
>>> query1 = iterator.filter('172.18.1.0/24').filter_key(['ipv4_network'])
>>> if query1.exists():
...   print(query1.first())
... 
Network(name=internal-network)
```

Let me know if that helps.

**Comment by gabstopper** [comment]
Hi, I wanted to let you know I pushed 0.5.2rc1 today and made one additional change to the search by filter. I initially implemented 'filter_key' as a mechanism to specify an attribute to get a real value from but decided to simplify that by using a kwarg on ``filter`` instead. 
Here is an example:

``
>>> list(Router.objects.filter(address='10.10.10.1'))
[Router(name=Router-10.10.10.1)]
``
It is an easier syntax to follow. 

The collection documentation has been improved quite a bit as well:
http://smc-python.readthedocs.io/en/latest/pages/collections.html

**Comment by gabstopper** [comment]
Hi Sebastien, Just wondering if it's ok to close this one. How are things going? Also, v0.5.5 released this week which supports the latest SMC version and many enhancements as well.

**Comment by gabstopper** [comment]
Hi sebastien, I haven't seen any activity on this for some time so i'll go ahead and close it. If you still have an issue, please re-open or open new case. Also, version 0.5.6 was just released with support for SMC 6.3.
